src/Parsers/shellParser.ts

```diff
@@ -0,0 +1,546 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
+ */
+
+import EventEmitter from 'events';
+import { Terminal } from 'xterm-headless';
+
+import logger from '../logging';
+import { SerialPort } from '../SerialPort/SerialPort';
+
+export interface ICallbacks {
+    onSuccess: (response: string, command: string) => void;
+    onError: (message: string, command: string) => void;
+    onTimeout?: (message: string, command: string) => void;
+}
+
+export type ShellParserSettings = {
+    shellPromptUart?: string;
+    logRegex: RegExp;
+    errorRegex: RegExp;
+    timeout: number;
+    columnWidth: number;
+};
+
+type CommandEnqueue = {
+    command: string;
+    callbacks: ICallbacks[];
+    sent: boolean;
+    sentTime: number;
+    timeout?: number;
+};
+
+export type ShellParser = Awaited<ReturnType<typeof shellParser>>;
+export type XTerminalShellParser = ReturnType<
+    typeof xTerminalShellParserWrapper
+>;
+
+export const xTerminalShellParserWrapper = (terminal: Terminal) => ({
+    getTerminalData: () => {
+        let out = '';
+        for (let i = 0; i <= terminal.buffer.active.cursorY; i += 1) {
+            const line = terminal.buffer.active.getLine(i);
+            if (typeof line !== 'undefined') {
+                out += `\r\n${line.translateToString()}`;
+            }
+        }
+
+        return out.trim();
+    },
+    clear: () => terminal.clear(),
+    getLastLine: () => {
+        const lastLine = terminal.buffer.active.getLine(
+            terminal.buffer.active.cursorY
+        );
+        if (typeof lastLine === 'undefined') {
+            return '';
+        }
+
+        return lastLine.translateToString().trim();
+    },
+    write: (data: string, callback: () => void | undefined) =>
+        terminal.write(data, callback),
+});
+
+export const shellParser = async (
+    serialPort: SerialPort,
+    xTerminalShellParser: XTerminalShellParser,
+    settings: ShellParserSettings = {
+        logRegex:
+            /[[][0-9]{2,}:[0-9]{2}:[0-9]{2}.[0-9]{3},[0-9]{3}] <([^<^>]+)> ([^:]+): .*(\r\n|\r|\n)$/,
+        errorRegex: /Error /,
+        timeout: 1000,
+        columnWidth: 80,
+    },
+    shellEchos = true
+) => {
+    const eventEmitter = new EventEmitter();
+
+    let commandBuffer = '';
+    let commandQueueCallbacks = new Map<string, ICallbacks[]>();
+    let commandQueue: CommandEnqueue[] = [];
+    let cr = false;
+    let crnl = false;
+    let pausedState = true; // Assume we have some command in the shell that has user has started typing
```

Comment:
This is because shell retains a pending command after you disconnect. Hence when you reconnect to that device an unsent command might already be typed in the prompt